kv/kvserver: TestStoreRangeUpReplicate failed

[(kv/kvserver).TestStoreRangeUpReplicate failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1792603&tab=buildLog) on [master@5987886452da9925ab6df82c163d842f4f6f722f](https://github.com/cockroachdb/cockroach/commits/5987886452da9925ab6df82c163d842f4f6f722f):

```
I200306 17:29:53.635929 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:29:54.636986 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:29:55.638458 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:29:56.640369 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:29:57.642667 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:29:58.644598 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:29:59.645478 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:00.649796 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:01.655058 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:02.669536 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:03.676048 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:04.450577 141534 kv/kvserver/replica_command.go:1602  [replicate,s1,r17/1:/Table/2{1-2}] change replicas (add [(n2,s2):2LEARNER] remove []): existing descriptor r17:/Table/2{1-2} [(n1,s1):1, next=2, gen=0]
E200306 17:30:04.532812 141534 kv/kvserver/queue.go:1089  [replicate,s1,r17/1:/Table/2{1-2}] change replicas of r17 failed: batch timestamp 0.000000123,19546 must be after replica GC threshold 1583512165.656611490,0
I200306 17:30:04.677976 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:05.679783 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:06.681255 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:07.682990 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:08.685772 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:09.701234 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:10.702268 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:11.707070 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:12.709946 123801 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200306 17:30:13.712091 144481 util/stop/stopper.go:539  quiescing
I200306 17:30:13.712770 144480 util/stop/stopper.go:539  quiescing
I200306 17:30:13.713286 144482 util/stop/stopper.go:539  quiescing
I200306 17:30:13.797278 144479 util/stop/stopper.go:539  quiescing
W200306 17:30:13.807051 124523 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 3: EOF:
W200306 17:30:13.807891 124319 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 1: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200306 17:30:13.809174 124046 gossip/gossip.go:1518  [n2] no incoming or outgoing connections
W200306 17:30:13.809613 124396 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 2: EOF:
W200306 17:30:13.811453 124526 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 3: rpc error: code = Unavailable desc = transport is closing:
W200306 17:30:13.808170 124326 gossip/gossip.go:1518  [n3] no incoming or outgoing connections
W200306 17:30:13.814339 124433 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 2: EOF:
W200306 17:30:13.816190 124484 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 1: rpc error: code = Canceled desc = grpc: the client connection is closing:
I200306 17:30:13.826635 144479 util/stop/stopper.go:539  quiescing
I200306 17:30:13.834831 144479 util/stop/stopper.go:539  quiescing
I200306 17:30:13.841481 144479 util/stop/stopper.go:539  quiescing
--- FAIL: TestStoreRangeUpReplicate (67.77s)
    client_raft_test.go:1556: condition failed to evaluate within 45s: not fully initialized
        goroutine 123801 [running]:
        runtime/debug.Stack(0xc000cc1c00, 0x6d301e0, 0xc0003c1300)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x6e41c80, 0xc0015a7500, 0xc000cc1c00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:37 +0x87
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeUpReplicate(0xc0015a7500)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_raft_test.go:1556 +0x42e
        testing.tRunner(0xc0015a7500, 0x5f44cc8)
        	/usr/local/go/src/testing/testing.go:909 +0x19a
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:960 +0x652

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeUpReplicate PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeUpReplicate.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreRangeUpReplicate failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1797175&tab=buildLog) on [master@e0ca60fcad75e4e81f6ab1def8a2e3ccc8dfed93](https://github.com/cockroachdb/cockroach/commits/e0ca60fcad75e4e81f6ab1def8a2e3ccc8dfed93):

```
I200309 14:32:35.084932 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:36.088369 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:37.095834 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:38.097786 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:39.098705 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:40.106015 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:41.109637 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:42.113257 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:43.115507 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:44.118379 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:45.120315 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:46.123373 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:47.134869 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:48.136215 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:49.154076 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:50.165810 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:51.171471 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:52.174290 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:53.177170 125223 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200309 14:32:54.179226 148402 util/stop/stopper.go:539  quiescing
I200309 14:32:54.179921 148400 util/stop/stopper.go:539  quiescing
I200309 14:32:54.180499 148401 util/stop/stopper.go:539  quiescing
E200309 14:32:54.187042 145754 kv/kvserver/consistency_queue.go:135  [consistencyChecker,s1,r32/1:/{Table/36-Max}] batch timestamp 0.000000123,21626 must be after replica GC threshold 1583760727.931209622,0
E200309 14:32:54.187555 145754 kv/kvserver/queue.go:1089  [consistencyChecker,s1,r32/1:/{Table/36-Max}] batch timestamp 0.000000123,21626 must be after replica GC threshold 1583760727.931209622,0
W200309 14:32:54.216977 125724 kv/kvserver/store.go:1541  [s3,r1/3:/{Min-System/NodeL…}] could not gossip first range descriptor: node unavailable; try another peer
I200309 14:32:54.320191 148399 util/stop/stopper.go:539  quiescing
W200309 14:32:54.325093 125688 gossip/gossip.go:1518  [n3] no incoming or outgoing connections
W200309 14:32:54.325637 125355 gossip/gossip.go:1518  [n2] no incoming or outgoing connections
W200309 14:32:54.327411 125825 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 1: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200309 14:32:54.330000 125892 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 2: EOF:
W200309 14:32:54.334430 125775 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 1: EOF:
W200309 14:32:54.339008 125949 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 3: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200309 14:32:54.339555 125937 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 3: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200309 14:32:54.340888 125807 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 2: EOF:
I200309 14:32:54.346114 148399 util/stop/stopper.go:539  quiescing
I200309 14:32:54.348921 148399 util/stop/stopper.go:539  quiescing
I200309 14:32:54.356260 148399 util/stop/stopper.go:539  quiescing
--- FAIL: TestStoreRangeUpReplicate (66.06s)
    client_raft_test.go:1556: condition failed to evaluate within 45s: not fully initialized
        goroutine 125223 [running]:
        runtime/debug.Stack(0xc000b0fc00, 0x6dd5920, 0xc00427d020)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x6ee8ac0, 0xc00011e600, 0xc000b0fc00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:37 +0x87
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeUpReplicate(0xc00011e600)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_raft_test.go:1556 +0x42e
        testing.tRunner(0xc00011e600, 0x5fdbc08)
        	/usr/local/go/src/testing/testing.go:909 +0x19a
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:960 +0x652

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeUpReplicate PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeUpReplicate.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreRangeUpReplicate failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1800278&tab=buildLog) on [master@6b96af1fbc9a7e331a957edb00e0f2f95f2792d7](https://github.com/cockroachdb/cockroach/commits/6b96af1fbc9a7e331a957edb00e0f2f95f2792d7):

```
I200310 15:27:56.835130 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:27:57.838769 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:27:58.840534 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:27:59.841443 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:00.842535 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:01.843425 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:02.848246 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:03.849400 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:04.853479 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:05.855916 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:06.864481 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:07.865569 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:08.866635 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:09.867584 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:10.871981 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:11.873259 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:12.875877 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:13.880687 126065 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200310 15:28:14.885427 199212 util/stop/stopper.go:539  quiescing
I200310 15:28:14.885637 199213 util/stop/stopper.go:539  quiescing
I200310 15:28:14.886676 199214 util/stop/stopper.go:539  quiescing
W200310 15:28:14.893562 126298 kv/txn.go:590  [liveness-hb] failure aborting transaction: node unavailable; try another peer; abort caused by: result is ambiguous (server shutdown)
I200310 15:28:14.893948 126298 kv/kvserver/node_liveness.go:804  [liveness-hb] retrying liveness update after kvserver.errRetryLiveness: result is ambiguous (server shutdown)
W200310 15:28:14.895719 126298 kv/txn.go:590  [liveness-hb] failure aborting transaction: node unavailable; try another peer; abort caused by: node unavailable; try another peer
W200310 15:28:14.896007 126298 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: node unavailable; try another peer
I200310 15:28:14.980545 199349 util/stop/stopper.go:539  quiescing
W200310 15:28:14.984457 126707 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 3: rpc error: code = Unavailable desc = transport is closing:
W200310 15:28:14.987884 126297 gossip/gossip.go:1518  [n2] no incoming or outgoing connections
W200310 15:28:14.988437 126770 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 2: EOF:
W200310 15:28:14.989805 126863 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 2: EOF:
W200310 15:28:14.990769 126696 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 1: rpc error: code = Unavailable desc = transport is closing:
W200310 15:28:14.991142 126536 gossip/gossip.go:1518  [n3] no incoming or outgoing connections
W200310 15:28:14.992404 126702 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 1: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200310 15:28:14.993304 126710 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 3: rpc error: code = Unavailable desc = transport is closing:
I200310 15:28:14.994686 199349 util/stop/stopper.go:539  quiescing
I200310 15:28:14.997999 199349 util/stop/stopper.go:539  quiescing
I200310 15:28:15.005686 199349 util/stop/stopper.go:539  quiescing
--- FAIL: TestStoreRangeUpReplicate (186.77s)
    client_raft_test.go:1556: condition failed to evaluate within 45s: not fully initialized
        goroutine 126065 [running]:
        runtime/debug.Stack(0xc000887c00, 0x6dec640, 0xc0003b38e0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x6effc60, 0xc0013ec900, 0xc000887c00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:37 +0x87
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeUpReplicate(0xc0013ec900)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_raft_test.go:1556 +0x42e
        testing.tRunner(0xc0013ec900, 0x5feeaf0)
        	/usr/local/go/src/testing/testing.go:909 +0x19a
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:960 +0x652

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeUpReplicate PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeUpReplicate.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreRangeUpReplicate failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1801028&tab=buildLog) on [master@06004bb20156e6191f24a1004d39d7f4e89658f2](https://github.com/cockroachdb/cockroach/commits/06004bb20156e6191f24a1004d39d7f4e89658f2):

```
W200310 20:35:26.381108 126122 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18287 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18287 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:26.381720 126122 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18287 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:26.550820 126325 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18307 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18307 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:26.551800 126325 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18307 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:26.643865 126466 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18324 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18324 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:26.644914 126466 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18324 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:26.809948 126122 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18344 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18344 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:26.810880 126122 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18344 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.001708 126325 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18364 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18364 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.002253 126325 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18364 must be after replica GC threshold 1583871876.894691439,0
I200310 20:35:27.104609 125996 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
W200310 20:35:27.148547 126466 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18380 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18380 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.149093 126466 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18380 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.260397 126122 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18398 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18398 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.261702 126122 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18398 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.447617 126325 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18418 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18418 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.448118 126325 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18418 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.550447 126466 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18435 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18435 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.552686 126466 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18435 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.722366 126122 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18455 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18455 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.723267 126122 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18455 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.906679 126325 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18475 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18475 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:27.907190 126325 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18475 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:28.023008 126466 kv/txn.go:590  [liveness-hb] failure aborting transaction: batch timestamp 0.000000123,18492 must be after replica GC threshold 1583871876.894691439,0; abort caused by: batch timestamp 0.000000123,18492 must be after replica GC threshold 1583871876.894691439,0
W200310 20:35:28.023730 126466 kv/kvserver/node_liveness.go:471  [liveness-hb] failed node liveness heartbeat: batch timestamp 0.000000123,18492 must be after replica GC threshold 1583871876.894691439,0
I200310 20:35:28.106525 151498 util/stop/stopper.go:539  quiescing
I200310 20:35:28.107111 151497 util/stop/stopper.go:539  quiescing
I200310 20:35:28.107817 151499 util/stop/stopper.go:539  quiescing
W200310 20:35:28.199171 126440 kv/kvserver/store.go:1541  [s3,r1/3:/{Min-System/NodeL…}] could not gossip first range descriptor: node unavailable; try another peer
I200310 20:35:28.268454 151496 util/stop/stopper.go:539  quiescing
W200310 20:35:28.277416 126570 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 3: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200310 20:35:28.279302 126737 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 2: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200310 20:35:28.279540 126398 gossip/gossip.go:1518  [n3] no incoming or outgoing connections
W200310 20:35:28.285441 126178 gossip/gossip.go:1518  [n2] no incoming or outgoing connections
I200310 20:35:28.289214 151496 util/stop/stopper.go:539  quiescing
I200310 20:35:28.291560 151496 util/stop/stopper.go:539  quiescing
I200310 20:35:28.298972 151496 util/stop/stopper.go:539  quiescing
--- FAIL: TestStoreRangeUpReplicate (70.99s)
    client_raft_test.go:1556: condition failed to evaluate within 45s: not fully initialized
        goroutine 125996 [running]:
        runtime/debug.Stack(0xc000c59c00, 0x6de99a0, 0xc001b73120)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x6efcf60, 0xc000bf1700, 0xc000c59c00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:37 +0x87
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeUpReplicate(0xc000bf1700)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_raft_test.go:1556 +0x42e
        testing.tRunner(0xc000bf1700, 0x5ff1150)
        	/usr/local/go/src/testing/testing.go:909 +0x19a
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:960 +0x652

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeUpReplicate PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeUpReplicate.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreRangeUpReplicate failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1806549&tab=buildLog) on [master@0f220ff65716cf2e5408e3fa6db887d2eb687e36](https://github.com/cockroachdb/cockroach/commits/0f220ff65716cf2e5408e3fa6db887d2eb687e36):

```
I200313 07:16:44.274510 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:45.275760 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:46.277744 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:47.283147 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:48.285196 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:49.286430 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:50.289911 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:51.293329 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:52.294746 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:53.297991 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:54.305793 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:55.313961 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:56.320033 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:57.337968 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:58.340020 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:16:59.342644 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:17:00.347491 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:17:01.375308 125653 kv/kvserver/client_raft_test.go:1556  SucceedsSoon: not fully initialized
I200313 07:17:02.378388 151603 util/stop/stopper.go:539  quiescing
I200313 07:17:02.378602 151604 util/stop/stopper.go:539  quiescing
I200313 07:17:02.379336 151605 util/stop/stopper.go:539  quiescing
E200313 07:17:02.389420 149287 kv/kvserver/consistency_queue.go:135  [consistencyChecker,s1,r19/1:/Table/2{3-4}] batch timestamp 0.000000123,22623 must be after replica GC threshold 1584080176.660185345,0
E200313 07:17:02.390646 149287 kv/kvserver/queue.go:1089  [consistencyChecker,s1,r19/1:/Table/2{3-4}] batch timestamp 0.000000123,22623 must be after replica GC threshold 1584080176.660185345,0
W200313 07:17:02.447290 126120 kv/kvserver/store.go:1541  [s3,r1/2:/{Min-System/NodeL…}] could not gossip first range descriptor: node unavailable; try another peer
I200313 07:17:02.499620 151522 util/stop/stopper.go:539  quiescing
W200313 07:17:02.514466 126323 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 2: rpc error: code = Unavailable desc = transport is closing:
W200313 07:17:02.515916 126003 gossip/gossip.go:1513  [n2] no incoming or outgoing connections
W200313 07:17:02.527203 126305 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 3: EOF:
W200313 07:17:02.528061 126050 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 2: EOF:
W200313 07:17:02.531011 126161 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 1: EOF:
W200313 07:17:02.532914 126206 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 1: rpc error: code = Unavailable desc = transport is closing:
W200313 07:17:02.534031 126180 gossip/gossip.go:1513  [n3] no incoming or outgoing connections
W200313 07:17:02.534603 126286 kv/kvserver/raft_transport.go:637  while processing outgoing Raft queue to node 3: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200313 07:17:02.542400 125766 gossip/gossip.go:1513  [n1] no incoming or outgoing connections
I200313 07:17:02.544813 151522 util/stop/stopper.go:539  quiescing
I200313 07:17:02.547634 151522 util/stop/stopper.go:539  quiescing
I200313 07:17:02.555839 151522 util/stop/stopper.go:539  quiescing
--- FAIL: TestStoreRangeUpReplicate (65.43s)
    client_raft_test.go:1556: condition failed to evaluate within 45s: not fully initialized
        goroutine 125653 [running]:
        runtime/debug.Stack(0xc000c91c00, 0x6e0d700, 0xc004b90ca0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x6f214c0, 0xc0012d1000, 0xc000c91c00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:37 +0x87
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeUpReplicate(0xc0012d1000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_raft_test.go:1556 +0x42e
        testing.tRunner(0xc0012d1000, 0x6012568)
        	/usr/local/go/src/testing/testing.go:909 +0x19a
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:960 +0x652

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeUpReplicate PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeUpReplicate.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/client_lease_test.go
+++ b/pkg/kv/kvserver/client_lease_test.go
@@ -95,6 +95,7 @@ func TestStoreRangeLeaseSwitcheroo(t *testing.T) {
 	sc := kvserver.TestStoreConfig(nil)
 	sc.TestingKnobs.DisableMergeQueue = true
 	sc.EnableEpochRangeLeases = true
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{storeConfig: &sc}
 	defer mtc.Stop()
 	mtc.Start(t, 1)

--- a/pkg/kv/kvserver/client_merge_test.go
+++ b/pkg/kv/kvserver/client_merge_test.go
@@ -234,6 +234,7 @@ func mergeWithData(t *testing.T, retries int64) {
 	storeCfg := kvserver.TestStoreConfig(nil)
 	storeCfg.TestingKnobs.DisableReplicateQueue = true
 	storeCfg.TestingKnobs.DisableMergeQueue = true
+	storeCfg.Clock = nil // manual clock
 
 	// Maybe inject some retryable errors when the merge transaction commits.
 	var mtc *multiTestContext
@@ -438,7 +439,7 @@ func mergeCheckingTimestampCaches(t *testing.T, disjointLeaseholders bool) {
 			if err != nil {
 				return err
 			}
-			if !rhsRepl.OwnsValidLease(mtc.clock.Now()) {
+			if !rhsRepl.OwnsValidLease(mtc.clock().Now()) {
 				return errors.New("rhs store does not own valid lease for rhs range")
 			}
 			return nil
@@ -453,7 +454,7 @@ func mergeCheckingTimestampCaches(t *testing.T, disjointLeaseholders bool) {
 		t.Fatal(pErr)
 	}
 
-	readTS := mtc.clock.Now()
+	readTS := mtc.clock().Now()
 
 	// Simulate a read on the RHS from a node with a newer clock.
 	var ba roachpb.BatchRequest
@@ -476,7 +477,7 @@ func mergeCheckingTimestampCaches(t *testing.T, disjointLeaseholders bool) {
 	pushee := roachpb.MakeTransaction("pushee", rhsKey, roachpb.MinUserPriority, readTS, 0)
 	pusher := roachpb.MakeTransaction("pusher", rhsKey, roachpb.MaxUserPriority, readTS, 0)
 	ba = roachpb.BatchRequest{}
-	ba.Timestamp = mtc.clock.Now()
+	ba.Timestamp = mtc.clock().Now()
 	ba.RangeID = rhsDesc.RangeID
 	ba.Add(pushTxnArgs(&pusher, &pushee, roachpb.PUSH_ABORT))
 	if br, pErr := rhsStore.Send(ctx, ba); pErr != nil {
@@ -509,7 +510,7 @@ func mergeCheckingTimestampCaches(t *testing.T, disjointLeaseholders bool) {
 	// disjoint or not because disjoint leaseholders will lead to a loss of
 	// resolution in the timestamp cache. Either way though, the transaction
 	// should not be allowed to create its record.
-	hb, hbH := heartbeatArgs(&pushee, mtc.clock.Now())
+	hb, hbH := heartbeatArgs(&pushee, mtc.clock().Now())
 	ba = roachpb.BatchRequest{}
 	ba.Header = hbH
 	ba.RangeID = lhsDesc.RangeID
@@ -570,6 +571,7 @@ func TestStoreRangeMergeTimestampCacheCausality(t *testing.T) {
 	ctx := context.Background()
 	storeCfg := kvserver.TestStoreConfig(nil /* clock */)
 	storeCfg.TestingKnobs.DisableMergeQueue = true
+	storeCfg.Clock = nil // manual clock
 	mtc := &multiTestContext{storeConfig: &storeCfg}
 	var readTS hlc.Timestamp
 	rhsKey := roachpb.Key("c")
@@ -871,6 +873,7 @@ func TestStoreRangeMergeStats(t *testing.T) {
 	ctx := context.Background()
 	storeCfg := kvserver.TestStoreConfig(nil)
 	storeCfg.TestingKnobs.DisableMergeQueue = true
+	storeCfg.Clock = nil // manual clock
 	mtc := &multiTestContext{storeConfig: &storeCfg}
 	mtc.Start(t, 1)
 	defer mtc.Stop()
@@ -1324,6 +1327,7 @@ func TestStoreRangeMergeRHSLeaseExpiration(t *testing.T) {
 	storeCfg := kvserver.TestStoreConfig(nil)
 	storeCfg.TestingKnobs.DisableReplicateQueue = true
 	storeCfg.TestingKnobs.DisableMergeQueue = true
+	storeCfg.Clock = nil // manual clock
 
 	// The synchronization in this test is tricky. The merge transaction is
 	// controlled by the AdminMerge function and normally commits quite quickly,
@@ -1410,7 +1414,7 @@ func TestStoreRangeMergeRHSLeaseExpiration(t *testing.T) {
 	// the transaction being inadvertently aborted during its first attempt,
 	// which this test is not designed to handle. If the merge transaction did
 	// abort then the get requests could complete on r2 before the merge retried.
-	hb, hbH := heartbeatArgs(mergeTxn, mtc.clock.Now())
+	hb, hbH := heartbeatArgs(mergeTxn, mtc.clock().Now())
 	if _, pErr := kv.SendWrappedWith(ctx, mtc.stores[0].TestSender(), hbH, hb); pErr != nil {
 		t.Fatal(pErr)
 	}
@@ -1509,6 +1513,7 @@ func TestStoreRangeMergeConcurrentRequests(t *testing.T) {
 	storeCfg.TestingKnobs.DisableSplitQueue = true
 	storeCfg.TestingKnobs.DisableMergeQueue = true
 	storeCfg.TestingKnobs.DisableReplicateQueue = true
+	storeCfg.Clock = nil // manual clock
 
 	var mtc *multiTestContext
 	storeCfg.TestingKnobs.TestingResponseFilter = func(
@@ -1984,7 +1989,7 @@ func TestStoreRangeMergeSlowUnabandonedFollower_WithSplit(t *testing.T) {
 		if err != nil {
 			return err
 		}
-		if !rhsRepl.OwnsValidLease(mtc.clock.Now()) {
+		if !rhsRepl.OwnsValidLease(mtc.clock().Now()) {
 			return errors.New("rhs store does not own valid lease for rhs range")
 		}
 		return nil
@@ -2204,7 +2209,7 @@ func TestStoreRangeMergeAbandonedFollowersAutomaticallyGarbageCollected(t *testi
 		if err != nil {
 			return err
 		}
-		if !rhsRepl.OwnsValidLease(mtc.clock.Now()) {
+		if !rhsRepl.OwnsValidLease(mtc.clock().Now()) {
 			return errors.New("store2 does not own valid lease for rhs range")
 		}
 		return nil
@@ -3190,6 +3195,7 @@ func TestStoreRangeMergeDuringShutdown(t *testing.T) {
 	storeCfg.TestingKnobs.DisableSplitQueue = true
 	storeCfg.TestingKnobs.DisableMergeQueue = true
 	storeCfg.TestingKnobs.DisableReplicateQueue = true
+	storeCfg.Clock = nil // manual clock
 
 	// Install a filter that triggers a shutdown when stop is non-zero and the
 	// rhsDesc requests a new lease.
@@ -3291,7 +3297,7 @@ func TestMergeQueue(t *testing.T) {
 
 	mtc.storeConfig = &storeCfg
 	// Inject clock for manipulation in tests.
-	mtc.clock = clock
+	mtc.storeConfig.Clock = clock
 	mtc.Start(t, 2)
 	defer mtc.Stop()
 	mtc.initGossipNetwork() // needed for the non-collocated case's rebalancing to work

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -441,6 +441,7 @@ func TestFailedReplicaChange(t *testing.T) {
 	runFilter.Store(true)
 
 	sc := kvserver.TestStoreConfig(nil)
+	sc.Clock = nil // manual clock
 	sc.TestingKnobs.EvalKnobs.TestingEvalFilter = func(filterArgs storagebase.FilterArgs) *roachpb.Error {
 		if runFilter.Load().(bool) {
 			if et, ok := filterArgs.Req.(*roachpb.EndTxnRequest); ok && et.Commit {
@@ -1160,6 +1161,7 @@ func TestReplicateAfterRemoveAndSplit(t *testing.T) {
 	sc.TestingKnobs.DisableReplicaGCQueue = true
 	// Disable eager replica removal so we can manually remove the replica.
 	sc.TestingKnobs.DisableEagerReplicaRemoval = true
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{
 		storeConfig: &sc,
 		// This test was written before the multiTestContext started creating many
@@ -1263,6 +1265,7 @@ func TestRefreshPendingCommands(t *testing.T) {
 			// Disable periodic gossip tasks which can move the range 1 lease
 			// unexpectedly.
 			sc.TestingKnobs.DisablePeriodicGossips = true
+			sc.Clock = nil // manual clock
 			mtc := &multiTestContext{
 				storeConfig: &sc,
 				// This test was written before the multiTestContext started creating
@@ -1744,7 +1747,7 @@ func TestProgressWithDownNode(t *testing.T) {
 		testutils.SucceedsSoon(t, func() error {
 			values := []int64{}
 			for _, eng := range mtc.engines {
-				val, _, err := storage.MVCCGet(context.Background(), eng, roachpb.Key("a"), mtc.clock.Now(),
+				val, _, err := storage.MVCCGet(context.Background(), eng, roachpb.Key("a"), mtc.clock().Now(),
 					storage.MVCCGetOptions{})
 				if err != nil {
 					return err
@@ -1803,6 +1806,7 @@ func runReplicateRestartAfterTruncation(t *testing.T, removeBeforeTruncateAndReA
 	// RaftElectionTimeoutTicks and rangeLeaseActiveDuration). This test expects
 	// mtc.stores[0] to hold the range lease for range 1.
 	sc.RaftElectionTimeoutTicks = 1000000
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{
 		storeConfig: &sc,
 		// This test was written before the multiTestContext started creating many
@@ -1900,6 +1904,7 @@ func testReplicaAddRemove(t *testing.T, addFirst bool) {
 	// and run it manually when we're ready.
 	sc.TestingKnobs.DisableReplicaGCQueue = true
 	sc.TestingKnobs.DisableEagerReplicaRemoval = true
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{
 		storeConfig: &sc,
 		// This test was written before the multiTestContext started creating many
@@ -1915,7 +1920,7 @@ func testReplicaAddRemove(t *testing.T, addFirst bool) {
 		return func() error {
 			values := make([]int64, len(mtc.engines))
 			for i, eng := range mtc.engines {
-				val, _, err := storage.MVCCGet(context.Background(), eng, key, mtc.clock.Now(),
+				val, _, err := storage.MVCCGet(context.Background(), eng, key, mtc.clock().Now(),
 					storage.MVCCGetOptions{})
 				if err != nil {
 					return err
@@ -2143,7 +2148,7 @@ func TestQuotaPool(t *testing.T) {
 		value := bytes.Repeat([]byte("v"), (3*quota)/4)
 		var ba roachpb.BatchRequest
 		ba.Add(putArgs(key, value))
-		if err := ba.SetActiveTimestamp(mtc.clock.Now); err != nil {
+		if err := ba.SetActiveTimestamp(mtc.clock().Now); err != nil {
 			t.Fatal(err)
 		}
 		if _, pErr := leaderRepl.Send(ctx, ba); pErr != nil {
@@ -2164,7 +2169,7 @@ func TestQuotaPool(t *testing.T) {
 		go func() {
 			var ba roachpb.BatchRequest
 			ba.Add(putArgs(key, value))
-			if err := ba.SetActiveTimestamp(mtc.clock.Now); err != nil {
+			if err := ba.SetActiveTimestamp(mtc.clock().Now); err != nil {
 				ch <- roachpb.NewError(err)
 				return
 			}
@@ -2202,6 +2207,7 @@ func TestWedgedReplicaDetection(t *testing.T) {
 	const rangeID = 1
 
 	sc := kvserver.TestStoreConfig(nil)
+	sc.Clock = nil // manual clock
 	// Suppress timeout-based elections to avoid leadership changes in ways
 	// this test doesn't expect.
 	sc.RaftElectionTimeoutTicks = 100000
@@ -2257,7 +2263,7 @@ func TestWedgedReplicaDetection(t *testing.T) {
 	value := []byte("value")
 	var ba roachpb.BatchRequest
 	ba.Add(putArgs(key, value))
-	if err := ba.SetActiveTimestamp(mtc.clock.Now); err != nil {
+	if err := ba.SetActiveTimestamp(mtc.clock().Now); err != nil {
 		t.Fatal(err)
 	}
 	if _, pErr := leaderRepl.Send(ctx, ba); pErr != nil {
@@ -2609,6 +2615,7 @@ func TestRaftAfterRemoveRange(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	storeCfg := kvserver.TestStoreConfig(nil /* clock */)
 	storeCfg.TestingKnobs.DisableMergeQueue = true
+	storeCfg.Clock = nil // manual clock
 	mtc := &multiTestContext{
 		storeConfig: &storeCfg,
 	}
@@ -3099,6 +3106,7 @@ func TestReplicateRogueRemovedNode(t *testing.T) {
 	// Newly-started stores (including the "rogue" one) should not GC
 	// their replicas. We'll turn this back on when needed.
 	sc.TestingKnobs.DisableReplicaGCQueue = true
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{
 		storeConfig: &sc,
 		// This test was written before the multiTestContext started creating many
@@ -3678,6 +3686,7 @@ func TestTransferRaftLeadership(t *testing.T) {
 	// and remove a replica in the middle of the test. Disable the
 	// replication queue; we'll control replication manually.
 	sc.TestingKnobs.DisableReplicateQueue = true
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{
 		storeConfig: &sc,
 		// This test was written before the multiTestContext started creating many
@@ -4357,7 +4366,6 @@ func TestTracingDoesNotRaceWithCancelation(t *testing.T) {
 	sc.TestingKnobs.DisableMergeQueue = true
 	mtc := &multiTestContext{
 		storeConfig: &sc,
-		clock:       hlc.NewClock(hlc.UnixNano, time.Millisecond),
 	}
 	mtc.Start(t, 3)
 	defer mtc.Stop()
@@ -4428,7 +4436,6 @@ func TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic(t *testing
 		t.Skip("disabled on teeing engine")
 	}
 
-	clock := hlc.NewClock(hlc.UnixNano, time.Nanosecond)
 	stopper := stop.NewStopper()
 	ctx := context.Background()
 	defer stopper.Stop(ctx)
@@ -4460,7 +4467,6 @@ func TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic(t *testing
 	sc := kvserver.TestStoreConfig(nil)
 	mtc := &multiTestContext{
 		storeConfig:     &sc,
-		clock:           clock,
 		rpcTestingKnobs: knobs,
 	}
 

--- a/pkg/kv/kvserver/client_replica_gc_test.go
+++ b/pkg/kv/kvserver/client_replica_gc_test.go
@@ -154,6 +154,7 @@ func TestReplicaGCQueueDropReplicaGCOnScan(t *testing.T) {
 	mtc := &multiTestContext{}
 	cfg := kvserver.TestStoreConfig(nil)
 	cfg.TestingKnobs.DisableEagerReplicaRemoval = true
+	cfg.Clock = nil // manual clock
 	mtc.storeConfig = &cfg
 
 	defer mtc.Stop()

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -126,7 +126,8 @@ func TestRejectFutureCommand(t *testing.T) {
 
 	manual := hlc.NewManualClock(123)
 	clock := hlc.NewClock(manual.UnixNano, 100*time.Millisecond)
-	mtc := &multiTestContext{clock: clock}
+	sc := kvserver.TestStoreConfig(clock)
+	mtc := &multiTestContext{storeConfig: &sc}
 	defer mtc.Stop()
 	mtc.Start(t, 1)
 
@@ -525,6 +526,7 @@ func setupLeaseTransferTest(t *testing.T) *leaseTransferTest {
 	}
 
 	cfg := kvserver.TestStoreConfig(nil)
+	cfg.Clock = nil // manual clock
 	// Ensure the node liveness duration isn't too short. By default it is 900ms
 	// for TestStoreConfig().
 	cfg.RangeLeaseRaftElectionTimeoutMultiplier =
@@ -912,6 +914,7 @@ func TestRangeLimitTxnMaxTimestamp(t *testing.T) {
 	cfg := kvserver.TestStoreConfig(nil)
 	cfg.RangeLeaseRaftElectionTimeoutMultiplier =
 		float64((9 * time.Second) / cfg.RaftElectionTimeout())
+	cfg.Clock = nil // manual clock
 	mtc := &multiTestContext{}
 	mtc.storeConfig = &cfg
 	keyA := roachpb.Key("a")
@@ -995,6 +998,7 @@ func TestLeaseMetricsOnSplitAndTransfer(t *testing.T) {
 			}
 			return nil
 		}
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{
 		storeConfig: &sc,
 		// This test was written before the multiTestContext started creating many
@@ -1091,6 +1095,7 @@ func TestLeaseNotUsedAfterRestart(t *testing.T) {
 	ctx := context.Background()
 
 	sc := kvserver.TestStoreConfig(nil)
+	sc.Clock = nil // manual clock
 	var leaseAcquisitionTrap atomic.Value
 	// Disable the split queue so that no ranges are split. This makes it easy
 	// below to trap any lease request and infer that it refers to the range we're
@@ -1545,7 +1550,7 @@ func TestRangeInfo(t *testing.T) {
 			EndKey: roachpb.KeyMax,
 		},
 	}
-	txn := roachpb.MakeTransaction("test", roachpb.KeyMin, 1, mtc.clock.Now(), 0)
+	txn := roachpb.MakeTransaction("test", roachpb.KeyMin, 1, mtc.clock().Now(), 0)
 	h.Txn = &txn
 	reply, pErr = kv.SendWrappedWith(context.Background(), mtc.distSenders[0], h, &scanArgs)
 	if pErr != nil {
@@ -1947,7 +1952,7 @@ func TestLeaseTransferInSnapshotUpdatesTimestampCache(t *testing.T) {
 	// mark of the new leaseholder's timestamp cache. Amusingly, if the bug
 	// we're regression testing against here still existed, we would not have
 	// to do this.
-	hb, hbH := heartbeatArgs(txnOld.TestingCloneTxn(), mtc.clock.Now())
+	hb, hbH := heartbeatArgs(txnOld.TestingCloneTxn(), mtc.clock().Now())
 	if _, pErr := kv.SendWrappedWith(ctx, mtc.stores[0].TestSender(), hbH, hb); pErr != nil {
 		t.Fatal(pErr)
 	}

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -1409,6 +1409,7 @@ func runSetupSplitSnapshotRace(
 	// Disable the split delay mechanism, or it'll spend 10s going in circles.
 	// (We can't set it to zero as otherwise the default overrides us).
 	sc.RaftDelaySplitToSuppressSnapshotTicks = -1
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{storeConfig: &sc}
 	defer mtc.Stop()
 	mtc.Start(t, 6)

--- a/pkg/kv/kvserver/client_test.go
+++ b/pkg/kv/kvserver/client_test.go
@@ -254,7 +254,6 @@ type multiTestContext struct {
 	t           testing.TB
 	storeConfig *kvserver.StoreConfig
 	manualClock *hlc.ManualClock
-	clock       *hlc.Clock
 	rpcContext  *rpc.Context
 	// rpcTestingKnobs are optional configuration for the rpcContext.
 	rpcTestingKnobs rpc.ContextTestingKnobs
@@ -317,7 +316,6 @@ func (m *multiTestContext) Start(t testing.TB, numStores int) {
 		mCopy := *m
 		mCopy.storeConfig = nil
 		mCopy.clocks = nil
-		mCopy.clock = nil
 		mCopy.engines = nil
 		mCopy.engineStoppers = nil
 		mCopy.startWithSingleRange = false
@@ -343,18 +341,29 @@ func (m *multiTestContext) Start(t testing.TB, numStores int) {
 	m.gossips = make([]*gossip.Gossip, numStores)
 	m.nodeLivenesses = make([]*kvserver.NodeLiveness, numStores)
 
-	if m.manualClock == nil {
-		m.manualClock = hlc.NewManualClock(123)
+	if m.storeConfig != nil && m.storeConfig.Clock != nil {
+		require.Nil(t, m.manualClock, "can't use manual clock; storeConfig.Clock is set")
+		require.Empty(t, m.clocks, "can't populate .clocks; storeConfig.Clock is set")
+		m.clocks = []*hlc.Clock{m.storeConfig.Clock}
+	} else if len(m.clocks) == 0 {
+		if m.manualClock == nil {
+			m.manualClock = hlc.NewManualClock(123)
+		}
+		m.clocks = []*hlc.Clock{hlc.NewClock(m.manualClock.UnixNano, time.Nanosecond)}
 	}
-	if m.clock == nil {
-		m.clock = hlc.NewClock(m.manualClock.UnixNano, time.Nanosecond)
+
+	if m.storeConfig != nil {
+		// Either they're equal, or the left is initially nil (see the golf
+		// above).
+		m.storeConfig.Clock = m.clocks[0]
 	}
+
 	if m.transportStopper == nil {
 		m.transportStopper = stop.NewStopper()
 	}
 	st := cluster.MakeTestingClusterSettings()
 	if m.rpcContext == nil {
-		m.rpcContext = rpc.NewContextWithTestingKnobs(log.AmbientContext{Tracer: st.Tracer}, &base.Config{Insecure: true}, m.clock,
+		m.rpcContext = rpc.NewContextWithTestingKnobs(log.AmbientContext{Tracer: st.Tracer}, &base.Config{Insecure: true}, m.clock(),
 			m.transportStopper, st, m.rpcTestingKnobs)
 		// Ensure that tests using this test context and restart/shut down
 		// their servers do not inadvertently start talking to servers from
@@ -393,9 +402,18 @@ func (m *multiTestContext) Start(t testing.TB, numStores int) {
 	})
 }
 
+func (m *multiTestContext) clock() *hlc.Clock {
+	return m.clocks[0]
+}
+
 func (m *multiTestContext) Stop() {
 	done := make(chan struct{})
 	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				m.t.Errorf("mtc.Stop() panicked: %+v", r)
+			}
+		}()
 		m.mu.RLock()
 
 		// Quiesce everyone in parallel (before the transport stopper) to avoid
@@ -598,14 +616,18 @@ func (t *multiTestContextKVTransport) SendNext(
 			t.mtc.mu.RUnlock()
 			if leaseHolderStore == nil {
 				// The lease holder is known but down, so expire its lease.
-				t.mtc.advanceClock(ctx)
+				if t.mtc.manualClock != nil {
+					t.mtc.advanceClock(ctx)
+				}
 			}
 		} else {
 			// stores has the range, is *not* the lease holder, but the
 			// lease holder is not known; this can happen if the lease
 			// holder is removed from the group. Move the manual clock
 			// forward in an attempt to expire the lease.
-			t.mtc.advanceClock(ctx)
+			if t.mtc.manualClock != nil {
+				t.mtc.advanceClock(ctx)
+			}
 		}
 	}
 	t.setPending(rep.ReplicaID, false)
@@ -790,7 +812,7 @@ func (m *multiTestContext) addStore(idx int) {
 	if len(m.clocks) > idx {
 		clock = m.clocks[idx]
 	} else {
-		clock = m.clock
+		clock = m.storeConfig.Clock
 		m.clocks = append(m.clocks, clock)
 	}
 	var eng storage.Engine
@@ -1392,12 +1414,12 @@ func (m *multiTestContext) heartbeatLiveness(ctx context.Context, store int) err
 // the desired effect would be ambiguous.
 func (m *multiTestContext) advanceClock(ctx context.Context) {
 	for i, clock := range m.clocks {
-		if clock != m.clock {
+		if clock != m.clock() {
 			log.Fatalf(ctx, "clock at index %d is different from the shared clock", i)
 		}
 	}
 	m.manualClock.Increment(m.storeConfig.LeaseExpiration())
-	log.Infof(ctx, "test clock advanced to: %s", m.clock.Now())
+	log.Infof(ctx, "test clock advanced to: %s", m.clock().Now())
 }
 
 // getRaftLeader returns the replica that is the current raft leader for the

--- a/pkg/kv/kvserver/consistency_queue_test.go
+++ b/pkg/kv/kvserver/consistency_queue_test.go
@@ -44,6 +44,7 @@ import (
 func TestConsistencyQueueRequiresLive(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	sc := kvserver.TestStoreConfig(nil)
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{storeConfig: &sc}
 	defer mtc.Stop()
 	mtc.Start(t, 3)
@@ -55,7 +56,7 @@ func TestConsistencyQueueRequiresLive(t *testing.T) {
 
 	// Verify that queueing is immediately possible.
 	if shouldQ, priority := mtc.stores[0].ConsistencyQueueShouldQueue(
-		context.TODO(), mtc.clock.Now(), repl, config.NewSystemConfig(sc.DefaultZoneConfig)); !shouldQ {
+		context.TODO(), mtc.clock().Now(), repl, config.NewSystemConfig(sc.DefaultZoneConfig)); !shouldQ {
 		t.Fatalf("expected shouldQ true; got %t, %f", shouldQ, priority)
 	}
 
@@ -64,7 +65,7 @@ func TestConsistencyQueueRequiresLive(t *testing.T) {
 	mtc.advanceClock(context.TODO())
 
 	if shouldQ, priority := mtc.stores[0].ConsistencyQueueShouldQueue(
-		context.TODO(), mtc.clock.Now(), repl, config.NewSystemConfig(sc.DefaultZoneConfig)); shouldQ {
+		context.TODO(), mtc.clock().Now(), repl, config.NewSystemConfig(sc.DefaultZoneConfig)); shouldQ {
 		t.Fatalf("expected shouldQ false; got %t, %f", shouldQ, priority)
 	}
 }
@@ -176,6 +177,7 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	sc := kvserver.TestStoreConfig(nil)
+	sc.Clock = nil // manual clock
 	mtc := &multiTestContext{
 		storeConfig: &sc,
 		// This test was written before the multiTestContext started creating many

--- a/pkg/kv/kvserver/node_liveness_test.go
+++ b/pkg/kv/kvserver/node_liveness_test.go
@@ -224,7 +224,7 @@ func TestNodeHeartbeatCallback(t *testing.T) {
 
 	// Verify that last update time has been set for all nodes.
 	verifyUptimes := func() error {
-		expected := mtc.clock.Now()
+		expected := mtc.clock().Now()
 		for i, s := range mtc.stores {
 			uptm, err := s.ReadLastUpTimestamp(context.Background())
 			if err != nil {
@@ -506,7 +506,7 @@ func TestNodeLivenessGetLivenesses(t *testing.T) {
 
 	livenesses := mtc.nodeLivenesses[0].GetLivenesses()
 	actualLMapNodes := make(map[roachpb.NodeID]struct{})
-	originalExpiration := mtc.clock.PhysicalNow() + mtc.nodeLivenesses[0].GetLivenessThreshold().Nanoseconds()
+	originalExpiration := mtc.clock().PhysicalNow() + mtc.nodeLivenesses[0].GetLivenessThreshold().Nanoseconds()
 	for _, l := range livenesses {
 		if a, e := l.Epoch, int64(1); a != e {
 			t.Errorf("liveness record had epoch %d, wanted %d", a, e)
@@ -985,7 +985,7 @@ func TestNodeLivenessDecommissionAbsent(t *testing.T) {
 	if err := mtc.dbs[0].CPut(ctx, keys.NodeLivenessKey(goneNodeID), &storagepb.Liveness{
 		NodeID:     goneNodeID,
 		Epoch:      1,
-		Expiration: hlc.LegacyTimestamp(mtc.clock.Now()),
+		Expiration: hlc.LegacyTimestamp(mtc.clock().Now()),
 	}, nil); err != nil {
 		t.Fatal(err)
 	}

--- a/pkg/kv/kvserver/replica_rangefeed_test.go
+++ b/pkg/kv/kvserver/replica_rangefeed_test.go
@@ -80,6 +80,7 @@ func TestReplicaRangefeed(t *testing.T) {
 
 	ctx := context.Background()
 	sc := kvserver.TestStoreConfig(nil)
+	sc.Clock = nil // manual clock
 	kvserver.RangefeedEnabled.Override(&sc.Settings.SV, true)
 	mtc := &multiTestContext{
 		storeConfig: &sc,
@@ -102,9 +103,9 @@ func TestReplicaRangefeed(t *testing.T) {
 	rangeID := mtc.Store(0).LookupReplica(startKey).RangeID
 
 	// Insert a key before starting the rangefeeds.
-	initTime := mtc.clock.Now()
+	initTime := mtc.clock().Now()
 	mtc.manualClock.Increment(1)
-	ts1 := mtc.clock.Now()
+	ts1 := mtc.clock().Now()
 	incArgs := incrementArgs(roachpb.Key("b"), 9)
 	_, pErr := kv.SendWrappedWith(ctx, db, roachpb.Header{Timestamp: ts1}, incArgs)
 	if pErr != nil {
@@ -177,7 +178,7 @@ func TestReplicaRangefeed(t *testing.T) {
 
 	// Insert a key non-transactionally.
 	mtc.manualClock.Increment(1)
-	ts2 := mtc.clock.Now()
+	ts2 := mtc.clock().Now()
 	pArgs := putArgs(roachpb.Key("c"), []byte("val2"))
 	_, pErr = kv.SendWrappedWith(ctx, db, roachpb.Header{Timestamp: ts2}, pArgs)
 	if pErr != nil {
@@ -186,7 +187,7 @@ func TestReplicaRangefeed(t *testing.T) {
 
 	// Insert a second key transactionally.
 	mtc.manualClock.Increment(1)
-	ts3 := mtc.clock.Now()
+	ts3 := mtc.clock().Now()
 	if err := mtc.dbs[1].Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
 		txn.SetFixedTimestamp(ctx, ts3)
 		return txn.Put(ctx, roachpb.Key("m"), []byte("val3"))
@@ -200,7 +201,7 @@ func TestReplicaRangefeed(t *testing.T) {
 
 	// Update the originally incremented key non-transactionally.
 	mtc.manualClock.Increment(1)
-	ts4 := mtc.clock.Now()
+	ts4 := mtc.clock().Now()
 	_, pErr = kv.SendWrappedWith(ctx, db, roachpb.Header{Timestamp: ts4}, incArgs)
 	if pErr != nil {
 		t.Fatal(pErr)
@@ -208,7 +209,7 @@ func TestReplicaRangefeed(t *testing.T) {
 
 	// Update the originally incremented key transactionally.
 	mtc.manualClock.Increment(1)
-	ts5 := mtc.clock.Now()
+	ts5 := mtc.clock().Now()
 	if err := mtc.dbs[1].Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
 		txn.SetFixedTimestamp(ctx, ts5)
 		_, err := txn.Inc(ctx, incArgs.Key, 7)

